com.unity.netcode.gameobjects/Tests/Editor/Serialization/BitWriterTests.cs

```diff
@@ -0,0 +1,333 @@
+
+                writer.WriteByte(0b11111111);
+                Assert.AreEqual(0b11111111_00001010_10001011, *asInt);
+            }
+        }
+
+        [Test]
+        public unsafe void TestWritingMultipleBitsFromLongs()
+        {
+            var writer = new FastBufferWriter(4, Allocator.Temp);
+            using (writer)
+            {
+                int* asInt = (int*)writer.GetUnsafePtr();
+
+                Assert.AreEqual(0, *asInt);
+
+                Assert.IsTrue(writer.VerifyCanWrite(3));
+                using (var bitWriter = writer.EnterBitwiseContext())
+                {
+                    bitWriter.WriteBits(0b11111111UL, 1);
+                    Assert.AreEqual(0b1, *asInt);
+
+                    bitWriter.WriteBits(0b11111111UL, 1);
+                    Assert.AreEqual(0b11, *asInt);
+
+                    bitWriter.WriteBits(0b11111110UL, 2);
+                    Assert.AreEqual(0b1011, *asInt);
+
+                    bitWriter.WriteBits(0b11111000UL, 4);
+                    Assert.AreEqual(0b10001011, *asInt);
+
+                    bitWriter.WriteBits(0b11111010UL, 4);
+                    Assert.AreEqual(0b1010_10001011, *asInt);
+                }
+
+                Assert.AreEqual(2, writer.Position);
+                Assert.AreEqual(0b1010_10001011, *asInt);
+
+                writer.WriteByte(0b11111111);
+                Assert.AreEqual(0b11111111_00001010_10001011, *asInt);
+            }
+        }
+
+        [Test]
+        public unsafe void TestWritingMultipleBytesFromLongs([Range(1, 64)] int numBits)
+        {
+            var writer = new FastBufferWriter(sizeof(ulong), Allocator.Temp);
+            using (writer)
+            {
+                ulong* asUlong = (ulong*)writer.GetUnsafePtr();
+
+                Assert.AreEqual(0, *asUlong);
+                var mask = 0UL;
+                for (var i = 0; i < numBits; ++i)
+                {
+                    mask |= (1UL << i);
+                }
+
+                ulong value = 0xFFFFFFFFFFFFFFFF;
+
+                Assert.IsTrue(writer.VerifyCanWrite(sizeof(ulong)));
+                using (var bitWriter = writer.EnterBitwiseContext())
+                {
+                    bitWriter.WriteBits(value, numBits);
+                }
+                Assert.AreEqual(value & mask, *asUlong);
+            }
+        }
+
+        [Test]
+        public unsafe void TestWritingBitsThrowsIfVerifyCanWriteNotCalled()
+        {
+            var writer = new FastBufferWriter(4, Allocator.Temp);
+            using (writer)
+            {
+                int* asInt = (int*)writer.GetUnsafePtr();
+
+                Assert.AreEqual(0, *asInt);
+
+                Assert.Throws<OverflowException>(() =>
+                {
+                    using (var bitWriter = writer.EnterBitwiseContext())
+                    {
+                        bitWriter.WriteBit(true);
+                    }
+                });
+
```

Comment:
Why are these all in the same test?

Reply:
The test is "verify trying to write without bounds checking throws an exception"... it's checking that for all the functions that should throw an exception. I would argue that this isn't four different tests of writing bits, it's one test of VerifyCanWriteBits

Reply:
I see, I think this is a difference in test strategy and a discussion for another time - I'm a proponent of one test per case, even though they're similar tests, they're different cases.